infra/Pos/Slotting/Types.hs

```diff
@@ -23,14 +33,56 @@ data EpochSlottingData = EpochSlottingData
 instance NFData EpochSlottingData
 
 -- | Data necessary for slotting to work which is basically part of GState.
-data SlottingData = SlottingData
-    { sdPenult      :: !EpochSlottingData
-    -- ^ 'EpochSlottingData' for penult (i. e. 'sdPenultEpoch') epoch.
-    , sdLast        :: !EpochSlottingData
-    -- ^ 'EpochSlottingData' for last epoch.
-    -- Last means 'sdPenultEpoch + 1'.
-    , sdPenultEpoch :: !EpochIndex
-    -- ^ Penult epoch for which 'EpochSlottingData' is known.
-    } deriving (Eq, Show, Generic)
+-- External code can use functions like getLastEpochIndex or getLastEpochSlottingData and 
+-- not worry about cases where it doesn't exist (this module should be responsible for it).
+-- Note that it's important to use error rather than default values like 0, because 
+-- such cases indicate invariants violation and shouldn't be hidden behind default values.
+type SlottingData = Map EpochIndex EpochSlottingData
+
+-- | Get the latest epoch index.
+getLastEpochIndex :: SlottingData -> EpochIndex
+getLastEpochIndex = fst . M.findMax
+
+-- | Get the latest epoch slotting data.
+getLastEpoch :: SlottingData -> EpochSlottingData
+getLastEpoch = snd . M.findMax
+
+-- | Get the penultimate epoch index. Last epoch - 1.
+getPenultEpochIndex :: SlottingData -> EpochIndex
+getPenultEpochIndex sd = decreaseEpochIndex $ getLastEpochIndex sd
+  where
+    decreaseEpochIndex :: EpochIndex -> EpochIndex
+    decreaseEpochIndex ei = EpochIndex $ getEpochIndex ei - 1
+
+-- | Get the penultimate epoch slotting data. Last epoch - 1.
+getPenultEpoch :: SlottingData -> EpochSlottingData
+getPenultEpoch sd = sd ! getPenultEpochIndex sd
+
+-- | Lookup the slotting data for an arbitrary `EpochIndex`.
+lookupEpochSlottingData :: EpochIndex -> SlottingData -> Maybe EpochSlottingData
+lookupEpochSlottingData = M.lookup
+
+-- | Insert `EpochSlottingData`.
+addEpochSlottingData :: EpochIndex -> EpochSlottingData -> SlottingData -> SlottingData
+addEpochSlottingData = M.insert
+
+computeSlotStart :: SlotId -> EpochSlottingData -> Timestamp
+computeSlotStart SlotId{..} esd = slotTimestamp siSlot esd
+    where
+      slotTimestamp (getSlotIndex -> locSlot) EpochSlottingData{..} =
+        esdStart + Timestamp (fromIntegral locSlot * convertUnit esdSlotDuration)
 
-instance NFData SlottingData
+findMatchingEpoch :: Monad m => Timestamp -> EpochIndex -> (EpochIndex -> m (Maybe EpochSlottingData)) -> m (Maybe (EpochIndex, EpochSlottingData))
+findMatchingEpoch approxCurrentTime lastEpochIndex getterEpochSlottingData = go lastEpochIndex
+  -- Iterate over lastEpochIndex..0 and find the epoch
```

Comment:
It can be done more efficiently using binary search, but I think it's not essential now :)